### ISSUE TITLE
Better per-class metric handling for unbalanced data sets

### ISSUE THREAD

**Issue by AnselmC** [PyTorchLightning/metrics#174]
## 🐛 Bug
For imbalanced data sets it is often desirable to compute metrics per-class; especially when performance w.r.t. the less prevalent class(es) is important and needs to be tracked.
However, such class imbalance also leads to the fact that a given mini batch will not necessarily contain samples of all classes.
If, for the same batch, the model being trained does not make predictions for every class either, computing a score for classes without predictions or targets makes no sense. Making sure that these values are not logged then requires special handling by the user.

#### Code sample
```python
import torch
import torchmetrics

f1 = torchmetrics.F1(num_lables=2, average="none")
preds = torch.tensor([0, 0, 0])
targets = torch.tensor([0, 0, 0])
f1(preds, targets) # returns tensor([1., 0.])
```
### Expected behavior
In the above example, computing f1 (or precision, recall, etc.) for class 1 only makes sense if class 1 is present in either the predictions or targets (or both, of course). Hence, I would expect the resulting tensor to encode this s.t. the user can more easily deal with this situation, e.g.
#### Code sample
```python
import torch
import torchmetrics

f1 = torchmetrics.F1(num_classes=2, average="none")
preds = torch.tensor([0, 0, 0])
targets = torch.tensor([0, 0, 0])
f1(preds, targets) # returns tensor([1., -1])
```

Unfortunately, encoding such missing scores with `-1` is not generally applicable, e.g. for Generalized IoU.
Maybe encoding the values as `inf` makes more sense...I presume this to be suboptimal as well though.
Another approach would be to have a possibly different `invalid`/`unknown` value per metric and expose this value via an attribute, i.e.

```python
f1 = torchmetrics.F1(num_classes=2, average="none")
f1.UNKNOWN_SCORE # -1
preds = torch.tensor([0, 0, 0])
targets = torch.tensor([0, 0, 0])
f1(preds, targets) # returns tensor([1., -1])
for class_id, score in enumerate(f1):
    if score == f1.UNKNOWN_SCORE:
        continue
    logger.log(f"f1_{class_id}", score)
```

### Environment
 - torchmetrics: 0.2.0

**Comment by Borda** [comment]
@AnselmC are you willing to work on this fix/enhancement?
cc: @PyTorchLightning/core-metrics

**Comment by AnselmC** [comment]
@Borda Yes, of course. I didn't immediately start because I wanted to get your input on this, i.e. in which direction to go.

**Comment by AnselmC** [comment]
@Borda should I create a draft PR for a single metric and then discuss based on this?

**Comment by Borda** [comment]
> @Borda should I create a draft PR for a single metric and then discuss based on this?

Yes, please 🦝
